frappe/core/doctype/user_permission/user_permission_list.js

```diff
@@ -48,5 +145,114 @@ frappe.listview_settings['User Permission'] = {
 
 			dialog.show();
 		});
+	},
+
+	validate: function(dialog, data) {
+		if(dialog.fields_dict.applicable_doctypes.get_unchecked_options().length == 0) {
+			data.apply_to_all_doctypes = 1;
+			data.applicable_doctypes = [];
+			return data;
+		}
+		if(data.apply_to_all_doctypes == 0 && !("applicable_doctypes" in data)) {
+			frappe.throw("Please select applicable Doctypes");
+		}
+		return data;
+	},
+
+	get_applicable_doctype: function(dialog) {
+		return new Promise(resolve => {
+			frappe.call({
+				method: 'frappe.core.doctype.user_permission.user_permission.check_applicable_doc_perm',
+				async: false,
+				args:{
+					user: dialog.fields_dict.user.value,
+					doctype: dialog.fields_dict.doctype.value,
+					docname: dialog.fields_dict.docname.value
+				}
+			}).then(r => {
+				resolve(r.message);
+			});
+		});
+	},
+
+	get_multi_select_options: function(dialog, applicable){
+		return new Promise(resolve => {
+			frappe.call({
+				method: 'frappe.desk.form.linked_with.get_linked_doctypes',
+				async: false,
+				args:{
+					user: dialog.fields_dict.user.value,
+					doctype: dialog.fields_dict.doctype.value,
+					docname: dialog.fields_dict.docname.value
+				}
+			}).then(r => {
+				var options = [];
+				for(var d in r.message){
+					var checked = ($.inArray(d, applicable) != -1) ? 1 : 0;
+					options.push({ "label":d, "value": d , "checked": checked});
+				}
+				resolve(options);
+			});
+		});
+	},
+
+	get_docname_options: function(dialog) {
+		return new Promise(resolve => {
```

Comment:
Instead of manually getting the docnames and setting the option... change the fieldtype of`docname` field to `Dynamic Link` and set its `Option` as `doctype`. It will efficiently pull the docname options for you
eg.

```javascript
{
                "fieldname": "docname",
		"fieldtype": "Dynamic Link",
		"options": "doctype",
}
```

Reply:
done